File: homework1/src/drawLetter.cpp
#include "ros/ros.h"
#include "geometry_msgs/Twist.h"
#include "turtlesim/Pose.h"
#include "turtlesim/SetPen.h"
#include "turtlesim/TeleportAbsolute.h"
#include <iostream>

using namespace std;

static int a = 0;
static double poseStart;
static double poseCurrent;

void poseCallback(const turtlesim::PoseConstPtr& msg) // Topic messages callback
{
  if (a==0) {
  	poseStart = msg->x;
  	a=1;
  }
  poseCurrent = msg->x;
  ROS_INFO("x: %.2f, y: %.2f", msg->x, msg->y);
}


int main(int argc, char ** argv)
{

    bool success = true;
    const double FORWARD_SPEED_MPS = 2.0;
    int16_t DISTANCE = 6;
    int16_t HORIZONTAL_DISTANCE = 3;
    
    string robot_name = string(argv[1]);

    ros::init(argc, argv, "move_turtle");   //Initialize the node
    ros::NodeHandle node;

    // сервис для мгновенной телепортации черепашки
    ros::service::waitForService("turtle1/teleport_absolute");
    ros::ServiceClient teleportClient = node.serviceClient<turtlesim::TeleportAbsolute>("turtle1/teleport_absolute");
    
    // сервис для отключения рисование пути и изменения его цвета
    ros::service::waitForService("turtle1/set_pen");
    ros::ServiceClient setpenClient = node.serviceClient<turtlesim::SetPen>("turtle1/set_pen");
    
    // сервис для очистки холста
    //ros::service::waitForService("clear");
    // ros::ServiceClient clearClient = node.serviceClient<std_srvs::Empty>("clear");


    // контейнер аргументов для вызова сервисов
    turtlesim::TeleportAbsolute teleport_arguments; 
    turtlesim::SetPen pen_arguments;

    // выключаем рисованиу пути, чтобы при телепортации черепашки не оставить след
    pen_arguments.request.off = 1;

    // выполняем запрос отключения рисования
    success = setpenClient.call(pen_arguments);

    if (!success) {
      ROS_ERROR_STREAM("Pen failed to switch off");
    }

    teleport_arguments.request.x = 3.0; // позиция по х
    teleport_arguments.request.y = 4.0; // позиция по у
    teleport_arguments.request.theta = 0; // 90 градусов

    // выполняем телепортацию черепашки
    success = teleportClient.call(teleport_arguments);
    
    if (!success) {
    ROS_ERROR_STREAM("Turtle failed to teleport" );
    }

    // Включаем рисование пути обратно и меняем цвет рисовки 
    pen_arguments.request.off = 0;
    pen_arguments.request.r = 240;
    pen_arguments.request.g = 45;
    pen_arguments.request.b = 240;
    pen_arguments.request.width = 16;
    success = setpenClient.call(pen_arguments);
    if (!success) {
    ROS_ERROR_STREAM("TurtlePen failed to switch on");
    }


    // A publisher for the movement data
    ros::Publisher pub = node.advertise<geometry_msgs::Twist>(robot_name + "/cmd_vel", 10);

    // A listener for pose
    ros::Subscriber sub = node.subscribe(robot_name + "/pose", 10, poseCallback);

    // Drive forward at a given speed. The robot points up the x-axis.
    // The default constuctor will set all commands to 0.
    geometry_msgs::Twist msg;
    msg.linear.y = FORWARD_SPEED_MPS;
    msg.angular.z = 0.0;

    // Loop at 10Hz, publishing movement conmmands until we shut down.
    ros::Rate rate(10);
    ROS_INFO("Starting to move forward");
    
    while (ros::ok())
    {

      // Левая сторона Н
      int16_t curDistance = 0;
      double t0 = ros::Time::now().toSec();
      cout << curDistance << endl;
      while (curDistance < DISTANCE) {
        double t1 = ros::Time::now().toSec();
        curDistance = abs((t1-t0)*msg.linear.y);
        pub.publish(msg);
      }
      ROS_INFO("STOP");

      // Возвращение к центру линии
      msg.linear.y = -FORWARD_SPEED_MPS;
      curDistance = 0;
      t0 = ros::Time::now().toSec();
      while (curDistance < DISTANCE / 2) {
        double t1 = ros::Time::now().toSec();
        curDistance = abs((t1-t0)*msg.linear.y);
        pub.publish(msg);
      }
      ROS_INFO("STOP");

      // Рисовка второго элемента Н
      msg.linear.y = 0;
      msg.linear.x = FORWARD_SPEED_MPS;
      curDistance = 0;
      t0 = ros::Time::now().toSec();
      while (curDistance < HORIZONTAL_DISTANCE) {
        double t1 = ros::Time::now().toSec();
        curDistance = abs((t1-t0)*msg.linear.x);
        pub.publish(msg);
      }

      // Рисовка верхней половины правой линии Н
      msg.linear.x = 0;
      msg.linear.y = FORWARD_SPEED_MPS;
      curDistance = 0;
      t0 = ros::Time::now().toSec();
      while (curDistance < DISTANCE / 2) {
        double t1 = ros::Time::now().toSec();
        curDistance = abs((t1-t0)*msg.linear.y);
        pub.publish(msg);
      }
      ROS_INFO("STOP");


      // Риcовка оставшейся нижней части
      msg.linear.y = -FORWARD_SPEED_MPS;
      curDistance = 0;
      t0 = ros::Time::now().toSec();
      while (curDistance < DISTANCE) {
        double t1 = ros::Time::now().toSec();
        curDistance = abs((t1-t0)*msg.linear.y);
        pub.publish(msg);
      }
      ROS_INFO("STOP");

      msg.linear.y = 0;

      pub.publish(msg);
      ros::spinOnce(); // Allow processing of incoming messages
      rate.sleep();
    }

    return 0;
}

File: homework1/scripts/draw_letter.py
#!/usr/bin/env python
import rospy
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
import sys
import math
from turtlesim.srv import TeleportAbsolute, SetPen


def pose_callback(pose):
    rospy.loginfo("Robot X=%f : Y=%f : Z=%f\n", pose.x, pose.y, pose.theta)
 
# сервис для мгновенной телепортации черепашки
def teleport_to(x,y,z):
    rospy.wait_for_service('/turtle1/teleport_absolute')
    try:
        teleport = rospy.ServiceProxy('/turtle1/teleport_absolute', TeleportAbsolute)
        teleport(x,y,z)
    except rospy.ServiceException as e:
        print("Service fail %s"%e) 

# сервис выключения рисования пути черепашки
def pen_off():
    rospy.wait_for_service('/turtle1/set_pen')
    try:
        pen_off = rospy.ServiceProxy('/turtle1/set_pen', SetPen)
        #       r   g   b   w    flag
        pen_off(210, 34, 45, 16, 255)
    except rospy.ServiceException as e:
        print("Pen service fail %s"%e)

# сервис включения рисования пути черепашки
def pen_on():
    rospy.wait_for_service('/turtle1/set_pen')
    try:
        pen_off = rospy.ServiceProxy('/turtle1/set_pen', SetPen)
        #       r  g  b   w  flag
        pen_off(210,34,45, 16, 0)
    except rospy.ServiceException as e:
        print("Pen service fail %s"%e)


def move_turtle(lin_vel,ang_vel):

    FORWARD_SPEED = 2

    rospy.init_node('draw_letter', anonymous=False)
    pub = rospy.Publisher('/turtle1/cmd_vel', Twist, queue_size=10)
    rospy.Subscriber('/turtle1/pose',Pose, pose_callback)

    rate = rospy.Rate(10) # 10hz
    angle = 2*math.pi*90/360 + 0.1
    distance = 6
    horizontal_distance = 3
    angle_speed = 30*2*math.pi/360
    vel = Twist()
    pen_off()
    teleport_to(3,4,0)
    pen_on()
    vel.linear.x = 0
    vel.linear.z = 0
    vel.angular.x = 0
    vel.angular.y = 0
    # vel.angular.z = abs(angle_speed)
    vel.angular.z = 0
    vel.linear.y = FORWARD_SPEED
    rospy.loginfo("LinearVel=%f: AngularVel=%f", lin_vel, ang_vel)
    while not rospy.is_shutdown():
        print("Start")
        # rotate 90 degrees
        # cur_angle = 0
        # t0 = rospy.Time().now().to_sec()
        # while(cur_angle < angle):
        #     t1=rospy.Time().now().to_sec()
        #     cur_angle = abs((t1-t0)*vel.angular.z)
        #     pub.publish(vel)
        # vel.linear.x = 2
        # vel.angular.z = 0

        # Левая сторона Н
        cur_distance = 0
        t0 = rospy.Time().now().to_sec()
        while(cur_distance < distance):
            t1 = rospy.Time().now().to_sec()
            cur_distance = abs((t1-t0)*vel.linear.y)
            pub.publish(vel)

        total = cur_distance
        cur_distance = 0
        vel.linear.y = -2
        
        # Возвращение к центру линии
        t0 = rospy.Time().now().to_sec()
        while(cur_distance < total/2):
            t1 = rospy.Time().now().to_sec()
            cur_distance = abs((t1-t0)*vel.linear.y)
            pub.publish(vel)
        
        vel.linear.y = 0
        vel.linear.x = 2
        
        # Рисовка второго элемента Н
        cur_distance = 0
        t0 = rospy.Time().now().to_sec()
        while(cur_distance < horizontal_distance):
            t1 = rospy.Time().now().to_sec()
            cur_distance = abs((t1-t0)*vel.linear.x)
            pub.publish(vel)
        
        vel.linear.x = 0
        vel.linear.y = FORWARD_SPEED
        
        cur_distance = 0
        # Рисовка верхней половины правой линии Н
        t0 = rospy.Time().now().to_sec()
        while(cur_distance < distance/2):
            t1 = rospy.Time().now().to_sec()
            cur_distance = abs((t1-t0)*vel.linear.y)
            pub.publish(vel)

        total = cur_distance
        cur_distance = 0
        vel.linear.x = 0
        vel.linear.y = -FORWARD_SPEED
        
        # Риовка оставшейся нижней части
        t0 = rospy.Time().now().to_sec()
        while(cur_distance < distance):
            t1 = rospy.Time().now().to_sec()
            cur_distance = abs((t1-t0)*vel.linear.y)
            pub.publish(vel)
        
        vel.linear.y = 0

        
        # vel.angular.z = 1
        # while True:
        #     vel.angular.z = 1

        vel.linear.y = 0
        pub.publish(vel)
        # rate.sleep()

if __name__ == '__main__':
    if True:
        print("Hello")
    try:
        move_turtle(float(sys.argv[1]),float(sys.argv[2]))
    except rospy.ROSInterruptException:
        pass